examples/dnn_dcgan_train_ex.cpp

```diff
@@ -0,0 +1,272 @@
+    }
+
+    // MNIST is broken into two parts, a training set of 60000 images and a test set of 10000
+    // images.  Each image is labeled so that we know what hand written digit is depicted.
+    // These next statements load the dataset into memory.
+    std::vector<matrix<unsigned char>> training_images;
+    std::vector<unsigned long>         training_labels;
+    std::vector<matrix<unsigned char>> testing_images;
+    std::vector<unsigned long>         testing_labels;
+    load_mnist_dataset(argv[1], training_images, training_labels, testing_images, testing_labels);
+
+    // Fix the random generator seeds for network initialization and noise
+    srand(1234);
+    dlib::rand rnd(std::rand());
+
+    // Instantiate both generator and discriminator
+    generator_type generator;
+    discriminator_type discriminator(
+        leaky_relu_(0.2f), leaky_relu_(0.2f), leaky_relu_(0.2f));
+    // Remove the bias learning from the networks
+    visit_layers(generator, visitor_no_bias());
+    visit_layers(discriminator, visitor_no_bias());
+    // Forward random noise so that we see the tensor size at each layer
+    discriminator(generate_image(generator, make_noise(rnd)));
+    cout << "generator" << endl;
+    cout << generator << endl;
+    cout << "discriminator" << endl;
+    cout << discriminator << endl;
+
+    // The solvers for the generator network.  In this case, we don't need to use a dnn_trainer
+    // for the generator, since we are going to train it manually
+    adam solver(0.f, 0.5f, 0.999f);
+    std::vector<adam> g_solvers(generator.num_computational_layers, solver);
+    double learning_rate = 2e-4;
+    // The discriminator trainer
+    dnn_trainer<discriminator_type, adam> d_trainer(discriminator, solver);
+    d_trainer.be_quiet();
+    d_trainer.set_learning_rate(learning_rate);
+    d_trainer.set_learning_rate_shrink_factor(1);
+    cout << d_trainer << endl;
+
+    // Resume training from last sync file
+    if (file_exists("dcgan_sync"))
+    {
+        deserialize("dcgan_sync") >> generator >> d_trainer;
+    }
+
+    const size_t minibatch_size = 64;
+    const std::vector<float> real_labels(minibatch_size, 1.f);
+    const std::vector<float> fake_labels(minibatch_size, -1.f);
+    dlib::image_window win;
+    size_t iteration = 0;
+    while (iteration < 50000)
+    {
+        // Train the discriminator with real images
+        std::vector<matrix<unsigned char>> real_samples;
+        while (real_samples.size() < minibatch_size)
+        {
+            auto idx = rnd.get_random_32bit_number() % training_images.size();
+            real_samples.push_back(training_images[idx]);
+        }
+        d_trainer.train_one_step(real_samples, real_labels);
+
+        // Train the discriminator with fake images
+        // 1. generate some random noise
+        std::vector<noise_t> noises;
+        while (noises.size() < minibatch_size)
+        {
+            noises.push_back(std::move(make_noise(rnd)));
```

Comment:
`std::move` not needed here since the return from `make_noise` is already an rvalue.  So it doesn't do anything.